File: 41-first-missing-positive/first-missing-positive.java
class Solution {
    public int firstMissingPositive(int[] A) {
        int max = Integer.MAX_VALUE;
        int n = A.length;
        boolean hasMaxAlready = false;
        for (int i=0; i<n; i++) {
            // edge case when length is Integer.MAX_VALUE
            if (A[i] == max) {
                hasMaxAlready = true;
            }
            // mark the non-contributing elements zero & negative values to Integer.MAX_VALUE
            if (A[i] <= 0) {
                A[i] = max;
            }
        }
        for (int i=0; i<n; i++) {
            int current = Math.abs(A[i]) - 1;
            // mark the indices negative
            if (current >= 0 && current < n && A[current] > 0) {
                A[current]  *= -1;
            }
        }
        for (int i=0; i<n; i++) {
            if (A[i] > 0) {
                return i+1;
            }
        }
        // edge case when the length is max
        if (!hasMaxAlready && n == max) {
            return max;
        }
        return n+1;
    }
}

File: 41-first-missing-positive/first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        one_present = 0
        for i in range(n):
            if nums[i] == 1:
                one_present = 1
            elif nums[i] <= 0 or nums[i] > n:
                nums[i] = 1
        if one_present == 0:
            return 1
        for i in range(n):
            current = abs(nums[i])
            if current == 1:
                continue
            elif nums[current-1] < 0:
                continue
            else:
                nums[current-1] *= -1
        for i in range(1, n):
            if nums[i] > 0:
                return i + 1
        return n + 1